File: mesh.ts
/**
 * Mesh Network for micro:bit
 * Extends radio range by relaying messages through intermediate devices
 *
 * Note: MakeCode radio has a fixed 19-byte limit per packet (not configurable like
 * MicroPython's radio.config(length=251)). We use binary format for max payload.
 *
 * Uses native C++ for packet build/parse on device; TypeScript fallback in simulator.
 */
//% weight=95 icon="\uf0ac" color=#E3008C
namespace mesh {

    const MESH_MAGIC = 0x4D;  // 'M'
    const HEADER_LENGTH = 6;   // magic + ttl + msgId(2) + srcId(2)
    const MAX_BUFFER_LENGTH = 19;  // MakeCode radio limit for sendBuffer
    const MAX_PAYLOAD_LENGTH = MAX_BUFFER_LENGTH - HEADER_LENGTH;  // 13 bytes

    let maxHops: number = 3;
    let relayEnabled: boolean = true;
    let messageId: number = 0;
    let seenIds: number[] = [];
    let seenPtr: number = 0;
    let onReceivedHandler: (message: string, signalStrength: number) => void = null;
    let initialized: boolean = false;

    function init() {
        if (initialized) return;
        initialized = true;
        radio.onReceivedBuffer(handleReceivedBuffer);
        seenIds = [];
        for (let i = 0; i < 16; i++) seenIds.push(-1);
    }

    function makeSeenKey(msgId: number, srcId: number): number {
        return (msgId << 16) | (srcId & 0xFFFF);
    }

    function wasSeen(key: number): boolean {
        for (let i = 0; i < 16; i++) {
            if (seenIds[i] === key) return true;
        }
        return false;
    }

    function markSeen(key: number) {
        seenIds[seenPtr] = key;
        seenPtr = (seenPtr + 1) % 16;
    }

    /**
     * Build mesh packet (C++ on device, TS in simulator)
     */
    //% shim=mesh::buildPacket
    function buildPacketNative(ttl: number, msgId: number, srcId: number, payload: Buffer): Buffer {
        // Simulator fallback
        const len = Math.min(payload.length, MAX_PAYLOAD_LENGTH);
        const packet = control.createBuffer(HEADER_LENGTH + len);
        packet[0] = MESH_MAGIC;
        packet[1] = ttl & 0xFF;
        packet.setNumber(NumberFormat.UInt16LE, 2, msgId & 0xFFFF);
        packet.setNumber(NumberFormat.UInt16LE, 4, srcId & 0xFFFF);
        packet.write(HEADER_LENGTH, payload.slice(0, len));
        return packet;
    }

    // Simulator state for parsed packet (getters read this when TS runs)
    let _parsedTtl = 0;
    let _parsedMsgId = 0;
    let _parsedSrcId = 0;
    let _parsedPayload: Buffer = null;

    /**
     * Parse mesh packet (C++ on device, TS in simulator)
     */
    //% shim=mesh::parsePacket
    function parsePacketNative(raw: Buffer): number {
        // Simulator fallback - store in module vars for getters
        if (!raw || raw.length < HEADER_LENGTH || raw[0] !== MESH_MAGIC) return 0;
        _parsedTtl = raw[1];
        _parsedMsgId = raw.getNumber(NumberFormat.UInt16LE, 2);
        _parsedSrcId = raw.getNumber(NumberFormat.UInt16LE, 4);
        _parsedPayload = raw.slice(HEADER_LENGTH);
        return 1;
    }

    //% shim=mesh::getParsedTtl
    function getParsedTtlNative(): number { return _parsedTtl; }
    //% shim=mesh::getParsedMsgId
    function getParsedMsgIdNative(): number { return _parsedMsgId; }
    //% shim=mesh::getParsedSrcId
    function getParsedSrcIdNative(): number { return _parsedSrcId; }
    //% shim=mesh::getParsedPayload
    function getParsedPayloadNative(): Buffer { return _parsedPayload; }

    function parseMeshPacket(buf: Buffer): { ttl: number, msgId: number, srcId: number, payload: string } {
        if (parsePacketNative(buf) !== 1) return null;
        const ttl = getParsedTtlNative();
        const msgId = getParsedMsgIdNative();
        const srcId = getParsedSrcIdNative();
        const payloadBuf = getParsedPayloadNative();
        let payload = "";
        if (payloadBuf) {
            try { payload = payloadBuf.toString(); } catch (e) { }
        }
        return { ttl, msgId, srcId, payload };
    }

    function buildMeshPacket(ttl: number, msgId: number, srcId: number, payload: string): Buffer {
        const buf = control.createBufferFromUTF8(payload);
        const len = Math.min(buf.length, MAX_PAYLOAD_LENGTH);
        const payloadBuf = buf.slice(0, len);
        return buildPacketNative(ttl, msgId, srcId, payloadBuf);
    }

    function truncateToPayloadLen(msg: string): string {
        const buf = control.createBufferFromUTF8(msg);
        if (buf.length <= MAX_PAYLOAD_LENGTH) return msg;
        let s = msg;
        while (control.createBufferFromUTF8(s).length > MAX_PAYLOAD_LENGTH) {
            s = s.substr(0, s.length - 1);
        }
        return s;
    }

    function handleReceivedBuffer(raw: Buffer) {
        const pkt = parseMeshPacket(raw);
        if (!pkt) return;

        const myId = control.deviceSerialNumber() & 0xFFFF;
        const key = makeSeenKey(pkt.msgId, pkt.srcId);

        if (pkt.srcId === myId) return;
        if (wasSeen(key)) return;
        markSeen(key);

        if (relayEnabled && pkt.ttl > 0) {
            const fwd = buildMeshPacket(pkt.ttl - 1, pkt.msgId, pkt.srcId, pkt.payload);
            radio.sendBuffer(fwd);
        }

        const signal = radio.receivedPacket(2);
        if (onReceivedHandler) {
            onReceivedHandler(pkt.payload, signal);
        }
    }

    /**
     * Send a message through the mesh network.
     * Message will be relayed by other micro:bits to extend range (max 13 characters).
     */
    //% block="mesh send message %message"
    //% blockId=mesh_send group="Send" weight=60
    export function sendMessage(message: string) {
        init();
        message = truncateToPayloadLen(message);
        const msgId = messageId;
        messageId = (messageId + 1) & 0xFFFF;
        const srcId = control.deviceSerialNumber() & 0xFFFF;
        const pkt = buildMeshPacket(maxHops, msgId, srcId, message);
        radio.sendBuffer(pkt);
    }

    /**
     * Run code when a mesh message is received (from any device in the network).
     */
    //% block="on mesh received"
    //% blockId=mesh_on_received blockGap=16 group="Receive" weight=55
    //% draggableParameters=reporter
    export function onReceived(handler: (message: string, signalStrength: number) => void) {
        init();
        onReceivedHandler = handler;
    }

    /**
     * Set the radio group (0-255). All micro:bits in the same group can communicate.
     */
    //% block="mesh set group %channel"
    //% blockId=mesh_set_group group="Config" weight=40
    //% channel.min=0 channel.max=255
    export function setGroup(channel: number) {
        init();
        radio.setGroup(Math.max(0, Math.min(255, channel)));
    }

    /**
     * Set maximum number of hops (relays). Higher = longer range, more network traffic.
     * @param hops Max hops (1-9). Default 3.
     */
    //% block="mesh set max hops %hops"
    //% blockId=mesh_set_max_hops group="Config" weight=39
    //% hops.min=1 hops.max=9 hops.defl=3
    export function setMaxHops(hops: number) {
        maxHops = Math.max(1, Math.min(9, hops));
    }

    /**
     * Enable or disable relaying received messages. Disable for leaf nodes to save power.
     */
    //% block="mesh set relay %enabled"
    //% blockId=mesh_set_relay group="Config" weight=38
    export function setRelay(enabled: boolean) {
        relayEnabled = enabled;
    }

    /**
     * Set radio transmit power (0-7). Higher = longer range, more power.
     */
    //% block="mesh set transmit power %power"
    //% blockId=mesh_set_power group="Config" weight=37 advanced=true
    //% power.min=0 power.max=7 power.defl=6
    export function setTransmitPower(power: number) {
        init();
        radio.setTransmitPower(Math.max(0, Math.min(7, power)));
    }
}

File: mesh.cpp
/**
 * Native C++ implementation for mesh network packet handling.
 * Builds and parses mesh packets - runs on micro:bit hardware.
 * Includes full mesh logic: parse, dedup, relay decision.
 * Simulator uses the TypeScript fallback in mesh.ts.
 */
#include "pxt.h"

using namespace pxt;

namespace mesh {

#define MESH_MAGIC 0x4D
#define HEADER_LENGTH 6
#define MAX_PAYLOAD_LENGTH 13
#define SEEN_SIZE 16

// Parsed packet state (used after parsePacket / processReceived)
static uint8_t s_ttl;
static uint16_t s_msgId;
static uint16_t s_srcId;
static uint8_t s_payloadBuf[MAX_PAYLOAD_LENGTH];
static int s_payloadLen;

// Mesh relay state
static bool s_shouldRelay = false;
static uint8_t s_relayPacket[HEADER_LENGTH + MAX_PAYLOAD_LENGTH];
static int s_relayPacketLen = 0;

// Deduplication
static uint32_t s_seenIds[SEEN_SIZE];
static int s_seenPtr = 0;
static bool s_relayEnabled = true;
static bool s_seenInitialized = false;

static void initSeenIds() {
    if (s_seenInitialized) return;
    for (int i = 0; i < SEEN_SIZE; i++)
        s_seenIds[i] = 0xFFFFFFFF;
    s_seenInitialized = true;
}

static uint32_t makeSeenKey(uint16_t msgId, uint16_t srcId) {
    return ((uint32_t)msgId << 16) | (srcId & 0xFFFF);
}

static bool wasSeen(uint32_t key) {
    for (int i = 0; i < SEEN_SIZE; i++) {
        if (s_seenIds[i] == key) return true;
    }
    return false;
}

static void markSeen(uint32_t key) {
    s_seenIds[s_seenPtr] = key;
    s_seenPtr = (s_seenPtr + 1) % SEEN_SIZE;
}

/**
 * Set whether this node relays received packets.
 */
//%
void setRelayEnabled(bool enabled) {
    s_relayEnabled = enabled;
}

/**
 * Build a mesh packet buffer from header fields and payload.
 * @param ttl Time-to-live (hop count)
 * @param msgId Message ID
 * @param srcId Source device ID
 * @param payload Payload buffer (max 13 bytes used)
 * @returns New buffer with formatted packet, or NULL on error
 */
//%
Buffer buildPacket(int ttl, int msgId, int srcId, Buffer payload) {
    if (payload == NULL) return NULL;

    int plen = payload->length;
    if (plen > MAX_PAYLOAD_LENGTH) plen = MAX_PAYLOAD_LENGTH;

    int totalLen = HEADER_LENGTH + plen;
    uint8_t buf[HEADER_LENGTH + MAX_PAYLOAD_LENGTH];

    buf[0] = MESH_MAGIC;
    buf[1] = (uint8_t)(ttl & 0xFF);
    buf[2] = (uint8_t)(msgId & 0xFF);
    buf[3] = (uint8_t)((msgId >> 8) & 0xFF);
    buf[4] = (uint8_t)(srcId & 0xFF);
    buf[5] = (uint8_t)((srcId >> 8) & 0xFF);

    for (int i = 0; i < plen; i++)
        buf[HEADER_LENGTH + i] = payload->data[i];

    return mkBuffer(buf, totalLen);
}

/**
 * Parse a received mesh packet. Call getParsed* to retrieve fields.
 * @param raw Raw received buffer
 * @returns 1 if valid mesh packet, 0 otherwise
 */
//%
int parsePacket(Buffer raw) {
    if (raw == NULL || raw->length < HEADER_LENGTH || raw->data[0] != MESH_MAGIC)
        return 0;

    s_ttl = raw->data[1];
    s_msgId = (uint16_t)(raw->data[2] | (raw->data[3] << 8));
    s_srcId = (uint16_t)(raw->data[4] | (raw->data[5] << 8));
    s_payloadLen = raw->length - HEADER_LENGTH;
    if (s_payloadLen < 0) s_payloadLen = 0;
    if (s_payloadLen > MAX_PAYLOAD_LENGTH) s_payloadLen = MAX_PAYLOAD_LENGTH;

    for (int i = 0; i < s_payloadLen; i++)
        s_payloadBuf[i] = raw->data[HEADER_LENGTH + i];

    return 1;
}

/**
 * Get parsed TTL (call after parsePacket returns 1)
 */
//%
int getParsedTtl() {
    return s_ttl;
}

/**
 * Get parsed message ID (call after parsePacket returns 1)
 */
//%
int getParsedMsgId() {
    return s_msgId;
}

/**
 * Get parsed source ID (call after parsePacket returns 1)
 */
//%
int getParsedSrcId() {
    return s_srcId;
}

/**
 * Get parsed payload as buffer (call after parsePacket returns 1)
 */
//%
Buffer getParsedPayload() {
    return mkBuffer(s_payloadBuf, s_payloadLen);
}

/**
 * Process a received mesh packet: parse, dedup, relay decision.
 * Call getParsed* and shouldRelay/getRelayPacket after this returns 1.
 * @param raw Raw received buffer
 * @param myId This device's ID (lower 16 bits of serial number)
 * @returns 1 if packet should be delivered to user, 0 if dropped
 */
//%
int processReceived(Buffer raw, int myId) {
    if (raw == NULL || raw->length < HEADER_LENGTH || raw->data[0] != MESH_MAGIC)
        return 0;

    initSeenIds();

    s_ttl = raw->data[1];
    s_msgId = (uint16_t)(raw->data[2] | (raw->data[3] << 8));
    s_srcId = (uint16_t)(raw->data[4] | (raw->data[5] << 8));
    s_payloadLen = raw->length - HEADER_LENGTH;
    if (s_payloadLen < 0) s_payloadLen = 0;
    if (s_payloadLen > MAX_PAYLOAD_LENGTH) s_payloadLen = MAX_PAYLOAD_LENGTH;

    for (int i = 0; i < s_payloadLen; i++)
        s_payloadBuf[i] = raw->data[HEADER_LENGTH + i];

    // Don't process our own messages (echo)
    if (s_srcId == (uint16_t)(myId & 0xFFFF))
        return 0;

    uint32_t key = makeSeenKey(s_msgId, s_srcId);
    if (wasSeen(key))
        return 0;

    markSeen(key);

    // Relay decision
    s_shouldRelay = false;
    s_relayPacketLen = 0;
    if (s_relayEnabled && s_ttl > 0) {
        s_shouldRelay = true;
        uint8_t newTtl = s_ttl - 1;
        s_relayPacket[0] = MESH_MAGIC;
        s_relayPacket[1] = newTtl;
        s_relayPacket[2] = (uint8_t)(s_msgId & 0xFF);
        s_relayPacket[3] = (uint8_t)((s_msgId >> 8) & 0xFF);
        s_relayPacket[4] = (uint8_t)(s_srcId & 0xFF);
        s_relayPacket[5] = (uint8_t)((s_srcId >> 8) & 0xFF);
        for (int i = 0; i < s_payloadLen; i++)
            s_relayPacket[HEADER_LENGTH + i] = s_payloadBuf[i];
        s_relayPacketLen = HEADER_LENGTH + s_payloadLen;
    }

    return 1;
}

/**
 * Whether the last processed packet should be relayed (call after processReceived returns 1)
 */
//%
bool shouldRelay() {
    return s_shouldRelay;
}

/**
 * Get the relay packet to send (call after processReceived returns 1 and shouldRelay is true)
 */
//%
Buffer getRelayPacket() {
    if (s_relayPacketLen <= 0) return NULL;
    return mkBuffer(s_relayPacket, s_relayPacketLen);
}

} // namespace mesh
